File: src/main/java/java2d/demos/Paint/AnimVal.java
package java2d.demos.Paint;

final class AnimVal
{
    private float curval;
    private float lowval;
    private float highval;
    private float currate;
    private float lowrate;
    private float highrate;

    public AnimVal(float lowval, float highval, float lowrate, float highrate) {
        this.lowval = lowval;
        this.highval = highval;
        this.lowrate = lowrate;
        this.highrate = highrate;
        this.curval = randval(lowval, highval);
        this.currate = randval(lowrate, highrate);
    }

    public AnimVal(float lowval, float highval, float lowrate, float highrate, float pos) {
        this(lowval, highval, lowrate, highrate);
        set(pos);
    }

    private float randval(float low, float high) {
        return (float) (low + Math.random() * (high - low));
    }

    public float getFlt() {
        return curval;
    }

    public int getInt() {
        return (int) curval;
    }

    public void anim() {
        curval += currate;
        clip();
    }

    public void set(float val) {
        curval = val;
        clip();
    }

    public void clip() {
        if (curval > highval) {
            curval = highval - (curval - highval);
            if (curval < lowval) {
                curval = highval;
            }
            currate = -randval(lowrate, highrate);
        } else if (curval < lowval) {
            curval = lowval + (lowval - curval);
            if (curval > highval) {
                curval = lowval;
            }
            currate = randval(lowrate, highrate);
        }
    }

    public void newlimits(float lowval, float highval) {
        this.lowval = lowval;
        this.highval = highval;
        clip();
    }
}

File: src/main/java/java2d/demos/Paint/AnimVal.kt
package java2d.demos.Paint

internal class AnimVal(
    private var lowVal: Float,
    private var highVal: Float,
    private val lowRate: Float,
    private val highRate: Float
) {
    var value: Float = 0f
        private set

    private var curRate: Float = 0f

    val intValue: Int get() = value.toInt()

    init {
        this.value = random(lowVal, highVal)
        this.curRate = random(lowRate, highRate)
    }

    constructor(lowVal: Float, highVal: Float, lowRate: Float, highRate: Float, initial: Float)
        : this(lowVal, highVal, lowRate, highRate)
    {
        set(initial)
    }

    private fun random(low: Float, high: Float): Float {
        return (low + Math.random() * (high - low)).toFloat()
    }

    fun anim() {
        value += curRate
        clip()
    }

    fun set(value: Float) {
        this.value = value
        clip()
    }

    private fun clip() {
        if (value > highVal) {
            value = highVal - (value - highVal)
            if (value < lowVal) {
                value = highVal
            }
            curRate = -random(lowRate, highRate)
        } else if (value < lowVal) {
            value = lowVal + (lowVal - value)
            if (value > highVal) {
                value = lowVal
            }
            curRate = random(lowRate, highRate)
        }
    }

    fun newLimits(lowVal: Float, highVal: Float) {
        this.lowVal = lowVal
        this.highVal = highVal
        clip()
    }
}
